base/sysinfo.jl

```diff
@@ -322,4 +324,83 @@ windows_version
 
 const WINDOWS_VISTA_VER = v"6.0"
 
+"""
+    Sys.isexecutable(path::String)
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function isexecutable(path::String)
+    if iswindows()
+        return isfile(path)
+    else
+        # We use `access()` and `X_OK` to determine if a given path is
+        # executable by the current user.  `X_OK` comes from `unistd.h`.
+        X_OK = 0x01
+        ccall(:access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    end
+end
+isexecutable(path::AbstractString) = isexecutable(String(path))
+
+"""
+    Sys.which(program_name::String)
+
+Given a program name, searches the current `PATH` to find the first binary with
+the proper executable permissions that can be run, and returns an absolute
+path. Raises `ErrorException` if no such program is available.  If a path with
+a directory in it is passed in for `program_name`, that exact path is tested
+for executable permissions only (with `.exe` and `.com` extensions added on
+Windows platforms), no searching of `PATH` is performed.
+"""
+function which(program_name::String)
+    # Build a list of program names that we're going to try
+    program_names = String[]
+    if iswindows()
+        # If the file already has an extension, try that name first
+        if !isempty(splitext(program_name)[2])
+            push!(program_names, program_name)
+        end
+
+        # But also try appending .exe and .com`
+        for pe in (".exe", ".com")
+            push!(program_names, string(program_name, pe))
+        end
+    else
+        # On non-windows, we just always search for what we've been given
+        push!(program_names, program_name)
+    end
+
+    path_dirs = String[]
+    program_dirname = dirname(program_name)
+    if isempty(program_dirname)
+        # If we have been given just a program name (not a relative or absolute
+        # path) then we should search `PATH` for it here:
+        pathsep = iswindows() ? ';' : ':'
+        path_dirs = split(get(ENV, "PATH", ""), pathsep)
+
+        # On windows we always check the current directory as well
+        if iswindows()
+            pushfirst!(path_dirs, pwd())
+        end
+    else
+        # If we've been given a path that has a directory name in it, then we
+        # check to see if that path exists (with all the attendant `program_names`
+        # skullduggery windows requires).  Otherwise, we search the PATH.
+        push!(path_dirs, program_dirname)
+    end
+
+    for path_dir in path_dirs
+        for program_name in program_names
+            program_path = joinpath(path_dir, program_name)
+            # If we find something that matches our name and we can execute
+            if isexecutable(program_path)
+                return program_path
+            end
+        end
+    end
+
+    # If we couldn't find anything, complain
+    error("$program_name not found")
```

Comment:
This should probably be an `ArgumentError`.

Reply:
Ah, I see it originally was but Jameson said it shouldn't be. Just using `error` here seems unfortunately general.

Reply:
Yeah, but I don't have a good feel for what errors should be what, so in the absence of a better idea, I'm going to keep it as `error()` for now.